### PR TITLE
Allow 8 Validator Multinode Cluster to Run Indefinitely

### PR DESCRIPTION
**Pull request by rauljordan**
This is part of #1586 

---

# Description

**Write why you are making the changes in this pull request**

This PR gets our multinode runtime setup with 8 validators to continue indefinitely. This has been a major team effort over the past few weeks and is now packaged into a PR.

**Write a summary of the changes you are making**

The changes included in the PR are as follows:
- [x] Standardize channel buffer sizes across the repository into a single parameter constant called `params.BeaconConfig().DefaultBufferSize`
- [x] Improves logging throughout the beacon and validator node by reducing amount of hashes and other noisy items being printed out to the console and only giving important information
- [x] Disable signature verification for our testnet
- [x] Adds a reorg counter for prometheus
- [x] Applies LMD GHOST fork choice rule
- [x] Uses historical states to roll back state and to update FFG checkpoints after a new finalized/justified slot is found
- [x] Adds more spans to block processing, stategen, and regular sync
- [x] Uses a fixed seed for shuffling (for now)
- [x] Only applies fork choice in regular_sync when blocks are received via p2p - blocks that are received via RPC do not go through LMD GHOST
- [x] We run regular_sync handlers in go routines in order to not block a main thread in case an operation gets lagged
- [x] We close p2p streams on the sender and recipient stream after they are done (MAJOR BUG FIX)
- [x] We add a gRPC deadline to requests to ensure no request can take up more than SLOT_DURATION seconds from the validator-beacon node interactions
- [x] Disable `ComputeStateRoot` for testnet as we will not have sig verification

### CODE REVIEW

File: beacon-chain/blockchain/fork_choice.go [hunk 1]
```diff
@@ -106,21 +104,43 @@ func (c *ChainService) updateFFGCheckPts(state *pb.BeaconState) error {
...
+	head, err := c.lmdGhost(justifiedHead, postState, attestationTargets)
 	if err != nil {
-		return fmt.Errorf("could not tree hash incoming block: %v", err)
+		return fmt.Errorf("could not run fork choice: %v", err)
 	}
-	// TODO(#1307): Use LMD GHOST as the fork-choice rule for Ethereum Serenity.
-	// TODO(#674): Handle chain reorgs.
-	if err := c.beaconDB.UpdateChainHead(block, computedState); err != nil {
+
+	// TODO(99999): Confirming with Nishant state gen can return state of any head slot. Using postState as a stub.
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/fork_choice.go]: ??

File: beacon-chain/blockchain/fork_choice.go [hunk 2]
```diff
@@ -106,21 +104,43 @@ func (c *ChainService) updateFFGCheckPts(state *pb.BeaconState) error {
...
 	}
-	log.WithField("blockRoot", fmt.Sprintf("0x%x", h)).Info("Chain head block and state updated")
+	h, err := hashutil.HashBeaconBlock(head)
+	if err != nil {
+		return fmt.Errorf("could not hash head: %v", err)
+	}
+	log.WithField("headRoot", fmt.Sprintf("0x%x", h)).Info("Chain head block and state updated")
 
-	if err := c.saveHistoricalState(computedState); err != nil {
+	// TODO(99999): Using postState as a stub.
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/fork_choice.go]: ???

File: beacon-chain/attestation/service.go [hunk 3]
```diff
@@ -105,7 +105,7 @@ func (a *Service) LatestAttestation(ctx context.Context, index uint64) (*pb.Atte
 
 	// return error if validator has no attestation.
 	if _, exists := a.Store[pubKey]; !exists {
-		return nil, fmt.Errorf("validator index %d does not have an attestation", index)
+		return nil, nil
```
Comment (IvanTheGreatDev) [inline beacon-chain/attestation/service.go]: Can you fix this comment here? Since it's technically not returning an error

File: beacon-chain/attestation/service.go [hunk 4]
```diff
@@ -5,6 +5,8 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/prysmaticlabs/prysm/shared/params"
+
```
Comment (prestonvanloon) [inline beacon-chain/attestation/service.go]: Space

File: beacon-chain/attestation/service.go (hunk 3 above)
Comment (prestonvanloon) [reply beacon-chain/attestation/service.go]: Can you return a grpc not found error?

File: beacon-chain/blockchain/block_processing.go [hunk 5]
```diff
@@ -106,17 +107,38 @@ func (c *ChainService) ReceiveBlock(ctx context.Context, block *pb.BeaconBlock)
 		return nil, fmt.Errorf("could not execute state transition with block %v", err)
 	}
 
+	log.WithFields(logrus.Fields{
+		"slotNumber":     block.Slot - params.BeaconConfig().GenesisSlot,
+		"justifiedEpoch": beaconState.JustifiedEpoch - params.BeaconConfig().GenesisEpoch,
+		"finalizedEpoch": beaconState.FinalizedEpoch - params.BeaconConfig().GenesisEpoch,
+	}).Info(
+		"State transition complete")
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/block_processing.go]: Weird formatting here

File: beacon-chain/core/helpers/committee.go [hunk 6]
```diff
@@ -214,9 +213,10 @@ func Shuffling(
 	committeesPerEpoch := EpochCommitteeCount(activeCount)
 
 	// Convert slot to bytes and xor it with seed.
-	slotInBytes := make([]byte, 32)
-	binary.LittleEndian.PutUint64(slotInBytes, slot)
-	seed = bytesutil.ToBytes32(bytesutil.Xor(seed[:], slotInBytes))
+	//slotInBytes := make([]byte, 32)
+	//binary.LittleEndian.PutUint64(slotInBytes, slot)
+	//seed = bytesutil.ToBytes32(bytesutil.Xor(seed[:], slotInBytes))
+	//seed := bytesutil.ToBytes32([]byte("fixed"))
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/committee.go]: Revert these changes, @nishant mentioned it works with shuffling

File: validator/client/validator_propose.go [hunk 7]
```diff
@@ -107,14 +105,7 @@ func (v *validator) ProposeBlock(ctx context.Context, slot uint64) {
 	}
 
 	// 3. Compute state root transition from parent block to the new block.
-	resp, err := v.proposerClient.ComputeStateRoot(ctx, block)
```
Comment (prestonvanloon) [inline validator/client/validator_propose.go]: Can you add all of this logic back?

File: beacon-chain/attestation/service.go (hunk 3 above)
Comment (prestonvanloon) [reply beacon-chain/attestation/service.go]: nvm this isnt grpc

File: beacon-chain/blockchain/block_processing.go [hunk 8]
```diff
@@ -171,14 +192,20 @@ func (c *ChainService) runStateTransition(
 		if err := c.updateFFGCheckPts(beaconState); err != nil {
 			return nil, fmt.Errorf("could not update FFG checkpts: %v", err)
 		}
+		// Save Historical States
```
Comment (terenc3t) [inline beacon-chain/blockchain/block_processing.go]:
```suggestion
		// Save Historical States.
```

File: beacon-chain/blockchain/fork_choice.go [hunk 9]
```diff
@@ -105,24 +115,50 @@ func (c *ChainService) updateFFGCheckPts(state *pb.BeaconState) error {
...
 
-	if err := c.saveHistoricalState(computedState); err != nil {
-		log.Errorf("Could not save new historical state: %v", err)
+		// Only regenerate head state if there was a reorg.
+		postState, err = stategenerator.GenerateStateFromBlock(c.ctx, c.beaconDB, head.Slot)
+		if err != nil {
+			return fmt.Errorf("could not gen state: %v", err)
+		}
+
+		if postState.Slot != postState.Slot {
```
Comment (terenc3t) [inline beacon-chain/blockchain/fork_choice.go]: this if condition looks weird wouldn't they always be the same?

File: beacon-chain/sync/regular_sync.go [hunk 10]
```diff
@@ -535,6 +542,10 @@ func (rs *RegularSync) receiveAttestation(msg p2p.Message) error {
 		log.Errorf("Could not hash received attestation: %v", err)
 		return err
 	}
+	log.WithFields(logrus.Fields{
+		"blockRoot":      fmt.Sprintf("%#x", attestation.Data.BeaconBlockRootHash32),
+		"justifiedEpoch": attestation.Data.JustifiedEpoch - params.BeaconConfig().GenesisEpoch,
+	}).Info("Received an attestation!")
```
Comment (terenc3t) [inline beacon-chain/sync/regular_sync.go]:
```suggestion
	}).Info("Received an attestation")
```

File: validator/client/validator_propose.go (hunk 7 above)
Comment (terenc3t) [reply validator/client/validator_propose.go]: +1

File: beacon-chain/blockchain/fork_choice.go (hunk 9 above)
Comment (prestonvanloon) [reply beacon-chain/blockchain/fork_choice.go]: : looks like a search and replace causality where the parameter name was renamed to postState and we didnt have good enough test coverage for a test to fail

File: validator/client/validator_propose.go (hunk 7 above)
Comment (prestonvanloon) [reply validator/client/validator_propose.go]: Looks like we merged without adding this back. I'll revert those changes
